Added position offsets

kuniss opened the pull request:
@linkrope  please take  a look on that PR. 

I'm added the optional possibility of reporting position offset additional to line and column to allow error marker computations in the EAG (Epsilon) language server be based on offset, making the implementation there more concise and simpler.
 
The both reporting types (line/column with code annotation versa offset/line/column w/ code annotation) can be controlled via the "--offset" command line option. Please assess the option implementation in Settings and its application.

Doing that I have improved the error reporting at the Earley parser as it hat another format as usual. See the comment for the particular commit.

kuniss committed:
Use the same error format as in the anlyzer.

kuniss committed:
Added offset computation

required by Epsilon language server for setting
error marker for generator created errors.

kuniss committed:
Added --offset option

to let the language server control force error message
creation including offset positions.

kuniss committed:
Ignore build directory

where generated code may be created at.

kuniss reviewed (commented):
This is a note regarding commit https://github.com/linkrope/gamma/pull/4/commits/c85feef1ce79629f445c240f091b2f9f8cfaf9da .

Originally an Earley parser error was reported as follows:
```
error: /home/denis/runtime-EclipseApplication/epsilon-test/simple.eag:4:6 S <+ 'Hello World!': N>: .
                                                                        ^ affixform incorrect
```
The code pointer was at the wrong position due to the "error: " prefix. Furthermore, the all over format was not as usual. Usually, an error is reported in 2 lines, first the error explanation, second the error position with code and a position pointer.

This correction makes the format unique to the one used by the analyser for all other position annotated error messages. So after the correction the same error shows up as:

```
error: affixform incorrect
/home/denis/runtime-EclipseApplication/epsilon-test/simple.eag:4:6 S <+ 'Hello World!': N>: .
                                                                        ^
```
This, at the end, is also a needed fix for my offset position extension.

If the while PR is not taken over, this commit should at least be cherry picked as it is a real implementation error.

linkrope commented on src/io.d:
```diff
@@ -51,6 +53,7 @@ struct Input
             col = 0;
         }
         index_ += text[index_ .. $].stride;
+        ++offset;
```
Does this work? Have you tried it with CR LF?
For example, with running `unix2dos` on some example.

kuniss replied:
Should work because Xtext is already implemented multi line ending aware (because they are running on Mac, Linux, Windows).
They only take NL into consideration as line ending, but every character counts into offset. 
See https://github.com/eclipse/xtext-core/blob/0fa87d4ba010cd11e7e90c587208e19cc30d808d/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.java#L42

linkrope commented on src/io.d:
```diff
@@ -133,14 +138,47 @@ struct Position
+            writer.put(link);
+        }
+        else {
+            const link = format!"%s@%s:%s:%s"(name, offset, line, col);
```
Do you still need `line` and `col` with the offset format?
I hoped that `abc.eag@42` would be enough.

kuniss replied:
Originally, I thought, I would need them just to have redundant information (thinking, redundant information is never bad). But in fact, it is not really needed and can be omitted.
I will update the PR for that.

linkrope commented on src/epsilon/main.d:
```diff
@@ -38,6 +38,7 @@ void main(string[] args)
                     "slag", "Generate SLAG evaluator.", &slag,
                     "sweep", "Generate single-sweep evaluator.", &sweep,
                     "soag", "Generate SOAG evaluator.", &soag,
+                    "offset", "Additionally reports offset position, omits erroneous code output", &offset,
```
What does "omits erroneous code output" mean?
Is `^` the problem? Or multi-line output in general?

What happens, if you use `--offset` together with `--verbose`?
Is the `trace` output ignored, or is it also "erroneous"?

kuniss replied:
I only want to say that the erroneous code and the "^" pointer is omitted.
Do you have a better phrase?
... But may be this is not needed if we move to the option "--lsp". :smile:

linkrope commented on src/epsilon/earley.d:
```diff
@@ -347,9 +347,9 @@ void Parse(int Dom, int Affixform, ref int Tree, bool Def) @safe
+                import epsilon.analyzer : Error;
 
-                error!"%s affixform incorrect"(MSymBuf[CurSym].Pos);
+                Error(MSymBuf[CurSym].Pos, "affixform incorrect");
```
Is it just the format that's wrong? Then I would prefer just fixing the format:

    error!"affixform incorrect\n%s"(MSymBuf[CurSym].Pos);

Or do you also need the incremented error count?

kuniss replied:
Yes, just the format is wrong. At least, that's what I have encountered because my regex did not match the original Erarly parser error reporting.
... I was not aware of an incremented error counter! May be it is also needed to be incremented in that case? We should check.

kuniss replied:
Shouldn't the error counter be incremented for Earley parser errors too? Earley parsing is part of the analyzer phase... I would say yes. In that sense my first solution of calling `analyzer.Error()` here would be right!
Additionally, calling that analzer method ensures the right right format for all analyzer errors and establish one and only one point to change the error reporting format. Both points are important for the language server support interface.

linkrope replied:
If `earley` calls `analyzer.Error`, this would cause a cyclic dependency.
Currently, the `analyzer` checks the parse tree returned by `earley.Parse` to increment the error count.
For the moment, this should be enough.

linkrope commented on src/epsilon/analyzer.d:
```diff
@@ -918,15 +923,18 @@ void Analyse(Input input)
     {
         EAG.History |= EAG.analysed;
-        info!"%s grammar is valid"(EAG.BaseName);
+        if (!reportOffset)
```
This is more than what you would guess "report offset" to do.
Is this necessary? Then:
shouldn't we have an option like "lsp" that also suppresses all logging below `warn`?

kuniss replied:
Yes, you are right. But was a later decision on my side to skip this last concluded information.
It would help definitely help to get no logging below `warn`. 
And the `--lsp` option looks much nicer! I was already played with that idea too. But the PR was a first try and I knew you would review it. :smiley:

I can do the --lsp option. But what would need to be done the filter out all logging below `warn`? I did not saw a central point in the code to do this...

linkrope replied:
Setting the log levels is as simple as this:
https://github.com/linkrope/gamma/blob/master/src/epsilon/main.d#L62-L63

kuniss replied:
You mean, setting the log level for --lsp option would be similar to:
```
if (lsp)
    levels |= Level.warn;
```
If, of course, Settings.lsp is defined.

linkrope replied:
```
if (lsp)
    levels = Level.error | Level.warn;
```

linkrope commented on src/epsilon/analyzer.d:
```diff
@@ -918,15 +923,18 @@ void Analyse(Input input)
-            error!"errors in %s"(EAG.BaseName);
-        else
-            error!"errors";
+        if (!reportOffset) { // skip final message as language server intergration does not need them  
```
I asked myself whether the log level for this message should have been "info".
If this would help with the language server, then we should change it.

kuniss replied:
Yes, should be "info". It's just a summary at the end. I guess, it was added originally, to not oversee the errors in case a lot of warning are trailing. But for LSP it is a problem, because it sets an obvious marker at pos (1,1) that there are errors, what is obvious due to the other real markers.

linkrope commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ tmp
 *.lst
 *.o
 *.obj
+build
```
How did you got the `build` directory?

kuniss replied:
The `build` directory is the default directory for Gradle for (temporary) build artifacts. I use it sometimes for all kind of generated artifacts. In the Epsilon VS code extension it is the default directory for generated compiler artifacts. That way it does not get covered by the git change detection mechanism (e.g., when I'm trying it out at the Gamma project).

But may be, we should give the generation directory another default name, e.g. "gen" or "eag-gen". What do you think?
In the final version, the Epsilon/Gamma/EAG VS code extension is intended to be used at any project having the generator carrying with it (im Bauch haben). So, we need a setting option (which is already there, in fact) where the generated artifact has to be written to in the local project context.

linkrope commented on src/epsilon/analyzer.d:
```diff
@@ -3,6 +3,7 @@ module epsilon.analyzer;
 import EAG = epsilon.eag;
 import Earley = epsilon.earley;
 import epsilon.lexer : Lexer, Token;
+import epsilon.settings : Settings;
```
unused import?

kuniss replied:
Yes, your are right.
I first passed the Settings "object" as second parameter to the analyzer.Analyse() method and later restricted it to the offset option value. Then forgot to remove the import.

kuniss committed:
Use the same error format as in the anlyzer.

kuniss committed:
Report error summarization message as info.

This will be helpful later on for the VS code
extension's language server error parsing.

kuniss committed:
Added offset computation and --ls option

required by Epsilon language server for setting
error marker for generator created errors.

kuniss committed:
Report error and warn only if --ls is on.

kuniss committed:
Adjusted unit test after method signature change

in previous commit for Input struct.
Boolean lsSupport parameter was added.

kuniss committed:
Improved UndefPos handling, added unit test

for the previous commit which added offset position reporting.

kuniss commented:
@linkrope please see my latest update on that PR. I tried to implement what we have discussed. Any comment welcome. :smile: 

But, please do not accept the PR. I want to reopen it as a new PR after having rebased/reworked the commits at my branch.
This PR can be closed then when all is alright.

linkrope commented on src/epsilon/main.d:
```diff
@@ -61,6 +62,9 @@ void main(string[] args)
     {
         if (verbose)
             levels |= Level.trace;
+        else if (lsSupport)
```
Now, verbose disables language-server support. My guess is, it should be the other way around?
```
if (lsSupport)
    ...
else if (verbose)
    ...
```

kuniss replied:
The idea was: _verbose_ would be requested by a human user only, never by the Epsilon LS. This should imply to get line/column positions with markers (^) instead of offset based positions. 
The verbose option would not work either for the Epsilon LS, the output parsing will fail.

kuniss replied:
Does this make sense?

linkrope replied:
There are two options:
1. language-server support does not work together with other switches
Then the application should fail with an error message.
2. language-server support could be used together with, for example `--verbose`
Perhaps an IDE supports CTRL-Click only with language-server offsets.
Then your solution would result in verbose output with offset positions.
What' s your preference?

kuniss replied:
I think the other switches should work. Only the `--verbose` is problematic. I'm not sure, if I always can skip the output when parsing.
Could we ensure the verbose output is always going to stdout?
Because the errors with offset position are always going to stderr, so I could ignore the verbose output very easily.
But I have to check, how verbose output behaves for the error parsing.

kuniss replied:
... All output is going to stderr.
I have to check whether the output parsing is robust enough for that.

kuniss replied:
Output parsing is robust enough. :smiley: 
And with commit c3aee25 we can just omit the special log level adjustment for `lsSupport` at all. I will remove it.

linkrope commented on src/epsilon/main.d:
```diff
@@ -39,6 +39,7 @@ void main(string[] args)
                     "sweep", "Generate single-sweep evaluator.", &sweep,
                     "soag", "Generate SOAG evaluator.", &soag,
                     "output-directory", "Write compiled compiler to directory.", &outputDirectory,
+                    "ls", "reports offset positions for accompanying VS code extension's language server", &lsSupport,
```
"ls" is no good abbreviation. Use "language-server" instead: too long to type, but expressive in any script.

kuniss replied:
Agree. I will change it, the explanation too as you mentioned below for the related comment.

linkrope commented on src/epsilon/settings.d:
```diff
@@ -35,6 +35,9 @@ struct Settings
     /// generate SOAG evaluator
     bool soag;
 
+    /// report offset for positions instead of line and column (needed by the Epsilon language server) 
```
That's one of (currently) two things. Keep the comment more general and mention offsets as example:

    /// support the Epsilon language server, for example by reporting offsets instead of line and column

linkrope commented on src/io.d:
```diff
@@ -5,17 +5,27 @@ import std.stdio;
 Input read(string name)
 {
-    return read(name, File(name));
+    return read(name, File(name), false);
```
Reading `false` is not nice. `std.typecons` has a magical `Flag` to improve the situation. Define it like this:
```
Input read(string name, Flag!"languageServer" languageServer = No.languageServer)
```
Then you can call it like this:
```
read(arg, Yes.languageServer)
```
The flag can be used as a condition.

For `getopt` there are two options:
- you use `&languageServer` (the `Flag!"languageServer"`):
  then you have to call the application with `--language-server=yes`
- or you keep the `bool`:
  then you have to convert it like `const languageServer = lsSupport ? Yes.languageServer : No.languageServer;`

kuniss replied:
Why is 
```
Input read(string name, Flag!"languageServer" languageServer = No.languageServer)
```
better than
```
Input read(string name, bool lsSupport = false)
```
When I stay with the boolean option for `lsSupport` in `getops`?
(I got only now aware of default parameter values in D).

That way I could stay with the same parameter passing for invocations of that method and can omit passing `false` or `true` like in
```
return read(name, File(name), false);
```

linkrope commented on src/io.d:
```diff
@@ -32,10 +42,15 @@ struct Input
+
+    private bool lsSupport;
+
+    this(string name, const(char)[] text, bool lsSupport) @nogc nothrow
```
Oh: in `io` we should better be specific: "use offset" rather than "support language server"
```
this(string name, const(char)[] text, Flag!"useOffset" useOffset = No.useOffset) @nogc nothrow
```

kuniss replied:
Same here: Why isn't using 
```
this(string name, const(char)[] text, bool useOffset = false) @nogc nothrow
```
simpler, under the circumstance that the `useOffset` argument value `true` is never passed explicitly but as variable value only (from `lsSupport` option variable)?
(And `false` gets never passed explicitely)

... Just trying to understand and learn. :smile:

linkrope commented on src/io.d:
```diff
@@ -51,6 +66,7 @@ struct Input
         index_ += text[index_ .. $].stride;
+        ++offset;
         ++col;
         if (lineBreak)
```
I could even image
```
if (lineBreak && !useOffset)
```
Then even here you don't need an additional `offset`.

linkrope commented on src/io.d:
```diff
@@ -68,11 +84,16 @@ struct Input
     Position position() const pure @safe
     {
-        import std.algorithm : find;
+        if (lsSupport) {    
```
I like the idea that `line` is set to 0 and is never incremented in case of `useOffset`.
Then no change is required here.

kuniss replied:
I'm note sure in that. The special circumstance of `line==0` implying offset handling is too implicit here. When reading it later, it will be hard to figure out. The passing of zero makes it more explicit, no further documentation hints needed.
Furthermore, we could avoid to `find` the text for position marking, as it is useless in case of offset reporting.

linkrope commented on src/io.d:
```diff
@@ -149,6 +176,19 @@ unittest
     assert(position.toString == expected);
 }
 
+@("convert position to offset")
```
@("convert offset position to string")

kuniss replied:
Was too late in the evening to get it right. :smiley:

linkrope commented on src/io.d:
```diff
@@ -149,6 +176,19 @@ unittest
+{
+    import std.string : outdent, strip;
+
+    const position = Position("äöü.txt", 0, 43, "äöü or empty");
```
Keep "äöü", to be as close as possible to the previous test.
Or use "don't care".

linkrope commented on src/io.d:
```diff
@@ -112,26 +133,32 @@ struct Position
-        writer.put(text.take(col - 1).map!(c => (c == '\t') ? c : ' '));
-        writer.put('^');
+        import std.format : format;
+        if (line == 0) { // means, lsSupport == true 
```
D code is compact enough that `{` gets a line of its own :-)

linkrope commented on src/io.d:
```diff
@@ -112,26 +133,32 @@ struct Position
+        import std.format : format;
+        if (line == 0) { // means, lsSupport == true 
+            const link = format!"%s@%s"(name, col);
+            writer.put(link);
```
"separate declarations and statements by an empty line"

linkrope commented:
> @linkrope please see my latest update on that PR. I tried to implement what we have discussed. Any comment welcome. 
> 
> But, please do not accept the PR. I want to reopen it as a new PR after having rebased/reworked the commits at my branch.
> This PR can be closed then when all is alright.

Usually we use "squash and merge": all commits are squashed into a single one; no traces are left from the various attempts.

kuniss committed:
rewording of comments and opts explanation

reagrding language server support.

kuniss committed:
Apply std.typecons:Flag mechanism

with default parameter value to get rid of defining
read() method 4 times, as suggested by @linkrope.

kuniss committed:
Replace typecons:Flag param by bool with default value

Looks simpler to me, but needs to be discussed with @linkrope.

kuniss committed:
Get rid of separate offset field.

Can be covered by col field if line breaks are getting ignored.

kuniss committed:
Some unit test improvements.

linkrope commented on src/epsilon/lexer.d:
```diff
@@ -495,6 +495,6 @@ version (unittest)
         auto symbolTable = new SymbolTable;
 
-        return Fixture(Lexer(Input("name", text), symbolTable), symbolTable);
+        return Fixture(Lexer(Input("name", text, false), symbolTable), symbolTable);
```
`false` is the default value, so it can be removed here.
(Otherwise, such a `true` or `false` argument strongly suggests a `Flag` parameter.)

kuniss replied:
Of cource. Forgot it.

linkrope commented on src/io.d:
```diff
@@ -2,20 +2,21 @@ module io;
+import std.typecons : Flag;
 
-Input read(string name)
+Input read(string name, bool lsSupport = false)
```
Keep "ls support" out of the `io` module.
For the `io` module, it's all about "use offset".

kuniss replied:
Agree completely. Good to have a reviewer. :smile:

linkrope commented on src/io.d:
```diff
@@ -28,14 +29,25 @@ struct Input
         this.name = name;
         this.text = text;
+        this.useOffset = useOffset;
+        if (useOffset) {
```
`{` gets a line of its own.

kuniss replied:
The else `{` too, I guess.

linkrope commented on src/io.d:
```diff
@@ -68,11 +80,16 @@ struct Input
     Position position() const pure @safe
     {
-        import std.algorithm : find;
+        if (useOffset) {    
```
Now, you can undo this `if`.
The extra work for the `text` argument is negligible.

kuniss replied:
Maybe, the extra work on text argument is negligible. But the special offset mode becomes explicit. Without the `if` it will be implicit and hard to figure out when reading and therefore would asking for a explanation comment which could be avoided otherwise.
(I'm repeating my argumentation just in case you did not see it. I will not insist in that. Just say, "I prefer may way" :smiley:)

linkrope replied:
If you write `if ` here then you should have two different `Position` constructors:
```
if (useOffset)
{
    return Position(name, col);
}
else
{
    ...
    return Position(name, line, col, text[begin .. end]);
}
```
Is this worth the extra effort?

kuniss replied:
I agree now. The offset semantic should be explicit enough with the code and comments in the Input constructor:
```
    this(string name, const(char)[] text, bool useOffset = false) @nogc nothrow
    {
        ...
        this.useOffset = useOffset;
        if (useOffset) 
        {
            this.line = 0;
            this.col = 0;
        }
        else 
        {
            this.line = 1;
            this.col = 1;
        }
    }
```

linkrope reviewed (approved):
Only trivial changes left.

kuniss committed:
Removed passing a default paramter value.

kuniss committed:
Talking about useOffset only

not refering to language server support here anymore as suggested correctly by @linkrope.
Additionally, improved block formating.

kuniss commented:
Some discussion points are open, see above. I set all others to resolved.

linkrope commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ tmp
 *.lst
 *.o
 *.obj
+build
```
Shown as "No newline at end of file".
At least, add the newline.

kuniss replied:
(Strange, I cannot answer per file comment here anymore.)
(Found the orginal discussion threads now, superfluous comments removed here.)

_.gitignore_
Of course.

kuniss committed:
Get rid of EOL warning

kuniss committed:
Revert to original Position creation

not differentiating betwenn line/column and offset
mode as both can handled implcitely with the same
parameter passing as discussed with @linkrope.

kuniss committed:
Removed special log level tweaking for LS support

as it is not needed anymore with commit
https://github.com/linkrope/gamma/commit/c3aee25318e868362b2037a4bd0933e5bb04531d
